---
title: "EWAS and Cell Type Deconvolution"
author: "Group 2"
date: "2025-03-02"
output: html_document
---

Install packages. Type no if it ever asks a follow-up question about updating old packages.
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("minfi")
BiocManager::install("limma")
BiocManager::install("qvalue")
BiocManager::install("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
BiocManager::install("biomaRt")
BiocManager::install("missMethyl")
BiocManager::install("EnhancedVolcano")
BiocManager::install("FlowSorted.Blood.EPIC")

install.packages("ggpubr")
install.packages("tidyr")
```


Loading required packages.
```{r, echo=FALSE}
library(dplyr)
library(minfi)
library(limma)
library(qvalue)
library(EnhancedVolcano)
library(FlowSorted.Blood.EPIC)
library(tidyr)
library(ggpubr)
library(ggplot2)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(biomaRt)
library(missMethyl)
```

In this analysis, we will be performing immune cell type deconvolution, an epigenome wide association study (EWAS), and a gene ontology (GO) enrichment analysis to study how the epignome differs in immune response of patients with breast cancer when compared to healthy controls.

First we will load the data. The data we will be working with are beta values which are a measure of relative methylation of a CpG site across the genome. Betas values range from 0 to 1 and represent the proportion of how many cells had a methylated base for that probe site. We would expect 0’s and 1’s in a perfect experiment (0 for an unmethylated locus and 1 for a methylated locus), but the reality is that technical noise and other types of variation make it very uncommon to encounter either of those scenarios. More often, beta values lie somewhere between 0.1-0.9, and we see a bimodal distribution with peaks at either end of that range when beta values are plotted on one line.

For more information on how the data is collected and how the beta value is calculated, go to this link:
[Click here to learn more about beta values](https://life-epigenetics-methylprep.readthedocs-hosted.com/en/latest/docs/introduction/introduction.html)

```{r}
# Loading the beta values.
load("../Data/beta_values.RData")

# Loading the phenotype data.
load("../Data/phenotype_data.RData")

# Show the first 5 rows of the phenotype dataframe.
head(beta_values)

# Show the first 5 rows of the phenotype data frame.
head(phenotype_data)
```

```{r}
# Checking that the phenotype and beta value dataframes are in the same order.
FALSE %in% c(row.names(phenotype_data) == colnames(beta_values))
```

================================================================================
================================================================================
Cell Type Deconvolution
================================================================================
================================================================================

Cell type deconvolution is a computational method used to estimate the proportions of different cell types in a mixed cell population from bulk DNA methylation data. Here we will be using the Flow.Sorted.Blood.EPIC tool to estimate six immune cell proportions from our bulk blood samples: T lymphocytes (CD4+ and CD8+), B cells, monocytes, natural killer cells, and neutrophils. To read more on cell type deconvolution: https://github.com/immunomethylomics/FlowSorted.BloodExtended.EPIC.

```{r}
# IDOLOptimizedCpGs.compTable are the model coefficients so subsetting the beta values to the CpGs in the model.
IDOLOptimizedCpGs.compTable <- IDOLOptimizedCpGs.compTable[row.names(IDOLOptimizedCpGs.compTable) %in% intersect(row.names(IDOLOptimizedCpGs.compTable), row.names(beta_values)),]
beta_values2 <- beta_values[, colnames(beta_values) %in% row.names(IDOLOptimizedCpGs.compTable)]
beta_values2 <- beta_values[row.names(IDOLOptimizedCpGs.compTable),]

# Estimating the cell type proportions.
immune_cell_type_proportions <- projectCellType_CP(as.matrix(beta_values2), as.matrix(IDOLOptimizedCpGs.compTable))
immune_cell_type_proportions <- as.data.frame(immune_cell_type_proportions)
immune_cell_type_proportions$SampleID <- row.names(immune_cell_type_proportions)
```

Plotting functions to create the stacked bar plot.

```{r}
pivot_cell_type_proportions_long <- function(cell_types){
  
  cell_types_cfp <- cell_types %>% pivot_longer(values_to = "CellTypeProportion",
                                                            names_to = "CellType", "CD8T":"Neu", 
                                                            values_drop_na = TRUE)
  return(cell_types_cfp)
  
}

addSmallLegend <- function(myPlot, pointSize = 5, textSize = 10, spaceLegend = 0.5) {
  myPlot +
    guides(shape = guide_legend(override.aes = list(size = pointSize)),
           color = guide_legend(override.aes = list(size = pointSize))) +
    theme(legend.title = element_text(size = textSize), 
          legend.text  = element_text(size = textSize),
          legend.key.size = unit(spaceLegend, "lines"))
}

stacked_bar_plot <- function(data, fill, x_factor, y, legend_position){
  
  stacked_bar_plot <- ggplot(data = data, aes(fill = fill, 
                                              x = factor(x_factor), y = y)) + 
    geom_bar(position="stack", stat="identity") + 
    ylab("Cell Type Proportions") + labs(fill = "CellType") +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), 
          axis.title.x = element_text(size = 18), axis.text.y = element_text(size = 18),
          axis.title.y = element_text(size = 18), legend.text = element_text(size = 36),
          legend.title = element_text(size = 36), legend.position = legend_position)
  
  return(stacked_bar_plot)
}

create_cell_type_proportion_plot <- function(cell_type_proportions,
                                             y_axis_label, x_axis_label, legend_position){
  
  # Cell Type Proportion Plot
  cell_types <- pivot_cell_type_proportions_long(cell_type_proportions)
  cell_types_cfp <- data.frame(cell_types)
  
  tumor_stacked_bar_plot <- stacked_bar_plot(data = cell_types_cfp, 
                                             fill = cell_types_cfp$CellType, 
                                             x_factor = cell_types_cfp$SampleID,
                                             y = cell_types_cfp$CellTypeProportion, legend_position = legend_position) + 
                                             ylab(y_axis_label) +
                                             xlab(x_axis_label)
  
  stacked_bar_plot <- addSmallLegend(tumor_stacked_bar_plot)
  
  return(stacked_bar_plot)
}
```

Creating a stacked barplot for all of the samples.

```{r}
# Creating the stacked barplot.
create_cell_type_proportion_plot(cell_type_proportions=immune_cell_type_proportions,
                                 y_axis_label="Immune Cell Type Proportions", x_axis_label="Bulk Blood Samples",       legend_position="bottom")
```
Comparing the cell type proportions of the breast cancer patients with healthy controls.

```{r}
# Plot function for comparing case vs controls
Plot_Function <- function(df){
  
  plot <- ggplot(
                df,
                aes(x = CellType, y = CellTypeProportions), fill=Condition) +
          geom_boxplot(
                aes(fill = factor(Condition)),
                width = 0.5,) +
          stat_compare_means(
                aes(group = Condition),
                label = "p.format",
                method = "wilcox.test") +
          theme(
                axis.title = element_text(size= 14),
                axis.text = element_text(size= 10),
                panel.background = element_rect(fill = "white", ),
                panel.grid = element_line(color = "#F1F1F1"),
                panel.border = element_rect(linewidth= 0.6, fill= NA),
                legend.title = element_text())
  
  return(plot)
}
```


```{r}
# Adding the condition phenotype datat to the immune cell type proportion data frame.
immune_cell_type_proportions$Condition <- phenotype_data$Condition

# Pivoting the data to long format for plotting.
immune_cell_type_proportions_long <- immune_cell_type_proportions %>% pivot_longer(values_to = "CellTypeProportions",
                                                            names_to = "CellType", "CD8T":"Neu", 
                                                            values_drop_na = TRUE)

# Plotting the cell type proportions.
Plot_Function(df=immune_cell_type_proportions_long)
```

================================================================================
================================================================================
Epigenome Wide Association Study (EWAS)
================================================================================
================================================================================

An Epigenome-Wide Association Study (EWAS) is a research approach used to identify associations between epigenetic modifications and specific traits, diseases, or environmental exposures. It is similar in concept to a Genome-Wide Association Study (GWAS), but instead of looking at genetic variations (SNPs), it examines DNA methylation changes across the genome.

To perform the EWAS, we use limma's lmFit: a function that fits multiple linear models by weighted or generalized least squares for each CpG probe (ex. cg25383568). If a model for a given probe has a significant beta coefficient (different then the beta value) then the probe is said to be significantly associated with the trait of interest. To learn more about limma's lmFit: https://kasperdanielhansen.github.io/genbioconductor/html/limma.html.

First we need to convert the beta values into M values. M-values take the ratio of methylated and unmethylated intensities (or beta value) and apply a log transformation. M-values are unbounded, so they can be more difficult to interpret and compare. A value around 0 means that the methylated and unmethylated signals are close to equal. A positive M-value means that the methylated signal is higher, while a negative M-value means that the unmethylated signal is higher.

```{r}
# Convert to M values.
m_values <- logit2(beta_values)
```

Creating the model matrix. A model matrix in statistical modeling and machine learning is a structured numerical representation of a dataset, where categorical variables are encoded into numeric form to be used in regression or other modeling techniques.

```{r}
# Changing the condition data type to factor setting "unaffected" to 0 and "affected" to 1.
phenotype_data$Condition <- factor(phenotype_data$Condition, levels = c("unaffected", "affected"))

# Define the model matrix.
XX <- model.matrix(~., data= phenotype_data)
```

```{r}
# Perform the EWAS - this function requires a lot of RAM, if your computer does not have enough and you get the following error:

# Error: vector memory limit of X Gb reached

# Skip to the code block that contains the line: load("../Data/results_df.RData")
fit <- lmFit(
  object = m_values,
  design = XX
);

fit <- eBayes(fit);
```

After running the EWAS, we need to perform multiple comparisons testing. An Epigenome-Wide Association Study (EWAS) examines hundreds of thousands (or even millions) of epigenetic sites (e.g., CpG sites) across the genome. Since statistical tests are performed at each site, this leads to a high risk of false positives (Type I errors). Multiple comparisons testing is necessary to control for this error rate and ensure that significant results are truly meaningful.

We will be using q-values to perform multiple comparisons testing. A q-value is a measure of statistical significance that controls the False Discovery Rate (FDR) in multiple hypothesis testing. It represents the minimum FDR at which a given test result (e.g., a p-value) is considered significant. Therefore, a q-value of <= 0.05 is significantly associated with our affected group.

```{r}
# Q Values.
q.values <- qvalue(fit$p.value[, "Conditionaffected"])

# Showing a summary of the number of significant CpG sites.
summary(q.values)
```

```{r}
# Generate the results table.
results_df <- data.frame('beta' = fit$coefficients[, "Conditionaffected"],
                              'pVal' = q.values$pvalues,
                              'log10.pVal' = -log10(q.values$pvalues),
                              'qVal' = q.values$qvalues)

# Arrange the results table in ascending order of q-value.
results_df <- results_df %>% arrange(qVal)
```

```{r}
# Skip if your computer had enough RAM to run lmFit()
# Run if your computer did not have enough RAM to run lmFit()
load("../Data/results_df.RData")
```

```{r}
# Show the first 5 lines of the results data frame.
head(results_df)
```

Now we will create a volcano plot is a scatter plot used to visualize the results of high-throughput analyses. In the context of an EWAS, it shows CpGs that are either significantly hypermethylated (those with a positive beta coefficient) or significantly hypomethylated (those with a negative beta coefficient).

```{r}
# Creating the volcano plot.
volcano <- EnhancedVolcano(results_df, lab = row.names(results_df),
                           x = 'beta', y = 'qVal', ylim = c(0,10),
                           pCutoff = 0.05, ylab = bquote(~-Log[10]~adjusted~italic(Q)),
                           xlab = bquote("Beta Coefficient"),
                           legendLabels = c("NS","BC","Sig q-value",
                           "Sig q-value & abs(BC) > 1"), drawConnectors = TRUE,
                           widthConnectors = 0.75)
volcano
```

================================================================================
================================================================================
GO Term Analysis
================================================================================
================================================================================

The purpose of this module is to explore the biological significance of CpGs.

First, let's take the top 20 significant CpGs and split into hyper-methylated and
hypo-methylated.
(20 is just for the sake of time)
```{r}
# Get 20 CpGs and split into up and down-regulated
sig_cpgs = results_df[1:20,]
upreg_cpgs = subset(sig_cpgs, beta > 0)
downreg_cpgs = subset(sig_cpgs, beta < 0)

# Keep just the CpG names
dif_methy_cpgs_up = row.names(upreg_cpgs)
dif_methy_cpgs_down = row.names(downreg_cpgs)
```

We'll then take a look at what pathways our significant CpGs are likely involved in
Under the table, this gometh() function will find CpG locations on the chromosome,
map them to genes, map those genes to biological pathways, and then test for enrichment
```{r}
# Get GO terms and their significance
CpG_locations_up <- gometh(dif_methy_cpgs_up, array.type = "EPIC", collection = "GO", sig.genes = TRUE)
CpG_locations_down <- gometh(dif_methy_cpgs_down, array.type = "EPIC", collection = "GO", sig.genes = TRUE)

# Only keep the top 20 pathways
topGSA_up <- topGSA(CpG_locations_up, n = 5)
topGSA_down <- topGSA(CpG_locations_down, n = 5)
```

Let's look at the pathways associated with our hypermethylated CpGs
```{r}
View(topGSA_up)
```

Now let's look at the pathways associated with our hypomethylated CpGs
```{r}
View(topGSA_down)
```
Note: this is simply a tutorial to teach you the pipeline and process. Typically, 
using only 20 CpGs is not enough to do a proper GO term analysis.


















